pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,274 @@
+		return cluster, nil
+	}
+
+	provider, ready := getProviderAndReadyStatus(cluster)
+	if provider == noKEv2Provider {
+		// not a KEv2 cluster
+		return cluster, nil
+	}
+
+	if !ready {
+		logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", cluster.Name)
+		return cluster, nil
+	}
+
+	var lastRefreshTime string
+	if cluster.Annotations != nil {
+		lastRefreshTime = cluster.Annotations[lastRefreshTime]
+	}
+
+	providerRefreshInterval, err := getProviderRefreshInterval(provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	nextRefreshTime, err := getNextProviderRefreshTime(providerRefreshInterval, lastRefreshTime)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !time.Now().After(nextRefreshTime) {
+		return cluster, err
+	}
+
+	cluster, err = c.refreshClusterUpstreamSpec(cluster, provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	c.clusterEnqueueAfter(key, providerRefreshInterval)
+
+	return cluster, nil
+}
+
+// getProviderAndReadyStatus returns the managed cluster provider of the given
+// custer and whether it is ready to be refresh and synced.
+func getProviderAndReadyStatus(cluster *mgmtv3.Cluster) (string, bool) {
+	// for other cloud drivers, please edit HERE
+	switch {
+	case cluster.Spec.EKSConfig != nil:
+		if cluster.Status.EKSStatus.UpstreamSpec == nil {
+			logrus.Infof("initial upstream spec for cluster [%s] has not been set by cluster handler yet, skipping", cluster.Name)
+			return apimgmtv3.ClusterDriverEKS, false
+		}
+		return apimgmtv3.ClusterDriverEKS, true
+	default:
+		return noKEv2Provider, false
+	}
+}
+
+// getProviderRefreshInterval returns the duration that should pass between
+// refreshing a cluster created by the given cloud provider.
+func getProviderRefreshInterval(provider string) (time.Duration, error) {
+	providerCronSetting, err := settings.GetSettingByID(fmt.Sprintf(refreshCronSettingFormat, provider))
+	if err != nil {
+		return 0, err
+	}
+
+	schedule, err := cron.Parse(providerCronSetting.Get())
+	if err != nil {
+		return 0, err
+	}
+
+	next := schedule.Next(time.Now())
+	return schedule.Next(next).Sub(next), nil
+}
+
+// getNextProviderRefreshTIme returns the next time a KEv2 cluster is due to refresh given
+// its last refresh time and cloud provider.
+func getNextProviderRefreshTime(refreshInterval time.Duration, lastRefreshAnno string) (time.Time, error) {
+	if lastRefreshAnno == "" {
+		return time.Now(), nil
+	}
+
+	lastRefreshUnix, err := strconv.ParseInt(lastRefreshAnno, 10, 64)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("unable to parse last KEv2 refresh time [%s]: %v", lastRefreshAnno, err)
+	}
+
+	return time.Unix(lastRefreshUnix, 0).Add(refreshInterval), nil
+}
+
+func (c *clusterRefreshController) refreshClusterUpstreamSpec(cluster *mgmtv3.Cluster, cloudDriver string) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
```

Comment:
This is redundant, the check is already done when the cluster first enters the handler